vms/platformvm/deposit/deposit.go

```diff
@@ -0,0 +1,140 @@
+)
+
+const (
+	interestRateBase        = 365 * 24 * 60 * 60
+	interestRateDenominator = 1_000_000 * interestRateBase
+)
+
+var bigInterestRateDenominator = (&big.Int{}).SetInt64(interestRateDenominator)
+
+type Offer struct {
+	ID ids.ID
+
+	UnlockHalfPeriodDuration uint32 `serialize:"true"`
+	InterestRateNominator    uint64 `serialize:"true"`
+	Start                    uint64 `serialize:"true"`
+	End                      uint64 `serialize:"true"`
+	MinAmount                uint64 `serialize:"true"`
+	MinDuration              uint32 `serialize:"true"`
+	MaxDuration              uint32 `serialize:"true"`
+}
+
+// Sets offer id from its bytes hash
+func (o *Offer) SetID() error {
+	bytes, err := blocks.GenesisCodec.Marshal(blocks.Version, o)
+	if err != nil {
+		return err
+	}
+	o.ID = hashing.ComputeHash256Array(bytes)
+	return nil
+}
+
+// Time when this offer becomes active
+func (o *Offer) StartTime() time.Time {
+	return time.Unix(int64(o.Start), 0)
+}
+
+// Time when this offer becomes inactive
+func (o *Offer) EndTime() time.Time {
+	return time.Unix(int64(o.End), 0)
+}
+
+func (o *Offer) InterestRateFloat64() float64 {
+	return float64(o.InterestRateNominator) / float64(interestRateDenominator)
+}
+
+type Deposit struct {
+	DepositOfferID      ids.ID `serialize:"true"`
+	UnlockedAmount      uint64 `serialize:"true"`
+	ClaimedRewardAmount uint64 `serialize:"true"`
+	Start               uint64 `serialize:"true"`
+	Duration            uint32 `serialize:"true"`
+	Amount              uint64 `serialize:"true"`
+}
+
+func (d *Deposit) StartTime() time.Time {
+	return time.Unix(int64(d.Start), 0)
+}
+
+func (d *Deposit) IsExpired(
+	depositOffer *Offer,
+	timestamp uint64,
+) bool {
+	return d.Start+uint64(d.Duration)+uint64(depositOffer.UnlockHalfPeriodDuration) < timestamp
+}
+
+// Returns amount of tokens that can be unlocked from [deposit] at [unlockTime] (seconds).
+//
+// Precondition: all args are valid in conjunction.
+func UnlockableAmount(deposit *Deposit, offer *Offer, unlockTime uint64) uint64 {
+	unlockPeriodStart, err := math.Add64(
+		deposit.Start,
+		uint64(deposit.Duration-offer.UnlockHalfPeriodDuration),
+	)
+	if err != nil || unlockPeriodStart > unlockTime {
+		// if err (overflow), than unlockPeriodStart > unlockTime for sure
+		return 0
+	}
+
+	unlockPeriodDuration := uint64(offer.UnlockHalfPeriodDuration) * 2
+	passedUnlockPeriodDuration := math.Min(unlockTime-unlockPeriodStart, unlockPeriodDuration)
+
+	bigTotalUnlockableAmount := (&big.Int{}).SetUint64(deposit.Amount)
+	bigPassedUnlockPeriodDuration := (&big.Int{}).SetUint64(passedUnlockPeriodDuration)
+	bigUnlockPeriodDuration := (&big.Int{}).SetUint64(unlockPeriodDuration)
+
+	// totalUnlockableAmount := depositAmount * passedUnlockPeriodDuration / unlockPeriodDuration
+	bigTotalUnlockableAmount.Mul(bigTotalUnlockableAmount, bigPassedUnlockPeriodDuration)
+	bigTotalUnlockableAmount.Div(bigTotalUnlockableAmount, bigUnlockPeriodDuration)
```

Comment:
do we really need the big.Int part/conversion? does the multiplication cause the concerns? what if we apply the division first?

Reply:
yes, the reason is multiplication - we can get overflow here
applying division first could result in less precision